src/main/query/where.c

```diff
@@ -63,352 +63,265 @@ static int AerospikeQuery_Where_Add(AerospikeQuery *self, PyObject *py_ctx,
-            as_index_datatype op_data =
-                (as_index_datatype)PyLong_AsLong(py_op_data);
-            rc = AerospikeQuery_Where_Add(
-                self, py_arg1, op, op_data,
-                size > 2 ? PyTuple_GetItem(py_arg2, 2) : Py_None,
-                size > 3 ? PyTuple_GetItem(py_arg2, 3) : Py_None,
-                size > 4 ? PyTuple_GetItem(py_arg2, 4) : Py_None,
-                size > 5 ? PyLong_AsLong(PyTuple_GetItem(py_arg2, 5)) : 0);
-            /* Failed to add the predicate for some reason */
-            if (rc != 0) {
-                as_error_update(&err, AEROSPIKE_ERR_PARAM,
-                                "Failed to add predicate");
+    PyObject *py_predicate_type = PyTuple_GetItem(py_predicate, 0);
+    if (!py_predicate_type) {
+        as_error_update(&err, AEROSPIKE_ERR_CLIENT,
+                        PREDICATE_INVALID_ERROR_MSG2);
+        goto CLEANUP;
+    }
+    PyObject *py_index_datatype = PyTuple_GetItem(py_predicate, 1);
+    if (!py_index_datatype) {
+        as_error_update(&err, AEROSPIKE_ERR_CLIENT,
+                        PREDICATE_INVALID_ERROR_MSG2);
+        goto CLEANUP;
+    }
+
+    if (!PyLong_Check(py_predicate_type) || !PyLong_Check(py_index_datatype)) {
+        as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                        PREDICATE_INVALID_ERROR_MSG1);
+        goto CLEANUP;
+    }
+
+    as_predicate_type predicate_type =
+        (as_predicate_type)PyLong_AsLong(py_predicate_type);
+    if (PyErr_Occurred()) {
+        goto CLEANUP;
+    }
+
+    as_index_datatype index_datatype =
+        (as_index_datatype)PyLong_AsLong(py_index_datatype);
+    if (PyErr_Occurred()) {
+        goto CLEANUP;
+    }
+
+    PyObject *py_bin = NULL;
+    PyObject *py_val1 = NULL;
+    PyObject *py_val2 = NULL;
+    PyObject **py_optional_tuple_items[] = {&py_bin, &py_val1, &py_val2};
+
+    // Read optional tuple items
+    const Py_ssize_t FIRST_OPTIONAL_IDX = 2;
+    for (Py_ssize_t i = FIRST_OPTIONAL_IDX; i <= 4; i++) {
+        PyObject *py_tuple_item;
+        if (i <= size - 1) {
+            py_tuple_item = PyTuple_GetItem(py_predicate, i);
+            if (!py_tuple_item) {
                 goto CLEANUP;
             }
-            /* Incorrect predicate or index type */
         }
         else {
-            as_error_update(&err, AEROSPIKE_ERR_PARAM, "predicate is invalid.");
+            py_tuple_item = Py_None;
+        }
+        *(py_optional_tuple_items[i - FIRST_OPTIONAL_IDX]) = py_tuple_item;
+    }
+
+    as_index_type index_type;
+    if (size == 6) {
+        PyObject *py_index_type = PyTuple_GetItem(py_predicate, 5);
+        if (!py_index_type) {
+            goto CLEANUP;
+        }
+        if (!PyLong_Check(py_index_type)) {
+            as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                            PREDICATE_INVALID_ERROR_MSG1);
+            goto CLEANUP;
+        }
+        index_type = PyLong_AsLong(py_index_type);
+        if (PyErr_Occurred()) {
```

Comment:
I like the named optional variables, however I find this block hard to follow with all the index arithmetic, cases, and magic numbers (i <= 4, size == 6). Why move from the old code's ternary expressions like `size > 2 ? PyTuple_GetItem(py_arg2, 2) : Py_None,` to this? I think it was easier to understand, if you want named variables maybe you could do something like ... `PyObject *py_val2 = size > 2 ? PyTuple_GetItem(py_arg2, 2) : Py_None` Is it for extensibility? If you want to keep this the way it is I think a comment block at top about the general form of a predicate tuple, what its max size is, which elements are optional, etc would help anyone that comes back to this code later.

Reply:
I rewrote it this way because I wanted to check that `PyTuple_GetItem()` doesn't set the error indicator (i.e raise an exception). In the old way, none of that error checking was being done:

https://github.com/aerospike/aerospike-client-python/blob/17.0.0/src/main/query/where.c#L390-L396

I think it's better practice to check the error indicator after calling a C-API function: https://docs.python.org/3/c-api/exceptions.html. When I tried to port the client to PyPy, it would seg fault because the error indicator was set before calling another C-API function

> If the error is not handled or carefully propagated, additional calls into the Python/C API may not behave as intended and may fail in mysterious ways